### ISSUE TITLE
juxt function

### ISSUE THREAD

**Issue by boxxxie** [ramda/ramda#986]
```
var juxt = R.curry(function(fns, thing){
  return R.map(function(fn){
    return fn(thing);
  },fns);
});
```
I seem to need this function a lot (very useful for comparing different queries over the same data).

does ramda have room for this? (very similar to R.ap, but more general, and none of the funny business of R.ap where it does some nested mapping).

**Comment by buzzdecafe** [comment]
This seems to me to be *exactly* `ap`:

`R.ap(fns, [thing])`

**Comment by boxxxie** [comment]
i don't believe so.
ap looks something like
```javascript
var ap = R.curry(function(fns, things){
  return R.compose(
    R.unnest,
    R.map(function(fn){
      return R.map(fn, things);
    }))
  (fns);
});
```
output http://ramdajs.com/docs/#ap
```
[ 2, 4, 6, 4, 5, 6 ]
```

juxt on the other hand.
```javascript
R.map(juxt([R.multiply(2), R.add(3)]), [1,2,3]);
```
output
```
[ [ 2, 4 ], [ 4, 5 ], [ 6, 6 ] ]
```

**Comment by buzzdecafe** [comment]
i see. this is breadth first. i stand corrected.

**Comment by CrossEye** [comment]
Although this looks like a very good idea, I think we might want to work through a more substantial implementation that allows the functions to be of higher arity.  Think:

    juxt([R.add, R.subtract, R.multiply, R.divide], 12,, 4); //=> [16, 8, 48, 3]

But this would be much more complicated to curry correctly.  As far as I can see, we'd have to do some sort of manual currying, which is made much harder by having to deal with `R.__` in all cases manually.  Nonetheless, it seems the right thing to do.  Alternately, we might do this:

    juxt([R.add, R.subtract, R.multiply, R.divide], [12,, 4]); //=> [16, 8, 48, 3]

which would be easier to implement, but would complicate the API for the simplest, and presumably most common case of `juxt(fns, val)`.

**Comment by buzzdecafe** [comment]
```javascript
R.map(juxt([R.multiply(2), R.add(3)]), [1,2,3]) <=>  R.map(R.pipe(R.of, R.ap([R.multiply(2), R.add(3)])), [1, 2, 3])
```

a little more cruft, but not much more

**Comment by CrossEye** [comment]
Because of the comment from @buzzdecafe, I'd like to see a good argument for including the functions, which might be seen almost as syntactic sugar on `ap`.  I think the example above with `map` is a bit misleading.  The appropriate parallel is:

    juxt([R.multiply(2), R.add(3)]), 1); //=> [2, 4]
    ap([R.multiply(2), R.add(3)]), [1]); //=> [2, 4]

My earlier comment is also probably off.  I think if we're going to do this in a manner to support multiple arguments, we'll probably have to do the fixed-arity, `juxt([fn..], [arg..])`, version.  Although there is a good argument to be made for the variadic version _once the list of functions has been curried in_, it really makes no sense before then.  And of course, in this version, `ap` already handles it.

**Comment by boxxxie** [comment]
I think it may be valuable to not accept an array for the last argument, and let the user decide if they want to map or not.
if the user uses map(juxt) they can name it.
if you guys like the auto-mapping, that's ok, but the lower level/simpler option should be done first, and build the more complicated stuff using it.

juxt is in clojure, underscore-contrib, and probably in other languages/libs, lets not change it for ramda.

**Comment by CrossEye** [comment]
My version with an array would not be for mapping but so that you could use it with functions that aren't unary:

    juxt ([R.add, R.multiply], [2, 5]); //=> [7, 10]

Although we could do this -- with some difficulty, I imagine -- with individual arguments, we've been moving towards a goal of having no variadic functions at all in Ramda.

But the points-free version of `juxt` @buzzdecafe suggests above has the same API as your original function.

**Comment by boxxxie** [comment]
sorry, i misunderstood what you guys were talking about.
i think that if you want to have functions of multi-arrity, then this would be a bit complicated.
in my experience, i have never needed juxt to act like this.
juxt is supposed to work with map, so adding this requirenment will break it.

**Comment by CrossEye** [comment]
The [Clojure docs][ce] for its version of `juxt` mentions that "the returned fn takes a variable number of args" and gives this example:

```clojure
;; sort list of maps by multiple values
(sort-by (juxt :a :b) [{:a 1 :b 3} {:a 1 :b 2} {:a 2 :b 1}]
;;=> [{:a 1 :b 2} {:a 1 :b 3} {:a 2 :b 1}]
```

That makes sense to me.  If we wanted to map single-arity functions, it would look a little more complicated:

    map(juxt(multiply(2), add(3)), [[1], [2], [3]]); //=> [[2, 4], [4, 5], [6, 6]]

It's slightly more complex syntactically than mapping over `[1, 2, 3]`, but not badly so.

Note that we're discussing a similar issue in #989.

  [ce]: https://clojuredocs.org/clojure.core/juxt

**Comment by boxxxie** [comment]
wow, i didn't actually know clojure juxt was like that.
anyway, clojure code maps pretty nicely to ramda.

the rest of clojure's library follows this convention (with map, filter, other fns).
ramda doesn't really follow this convention, so i think that the juxt i presented works nicely with the rest of ramda.

it may be nice to have both.

**Comment by buzzdecafe** [comment]
closing without prejudice, reopen if you want to pursue. as noted above, this function is easily enough produced by composition `var juxt = R.useWith(R.ap, R.identity, R.of)`

**Comment by samueller** [comment]
Do I understand this correctly in that the difference between using `juxt` and `ap` is that `ap` requires the functions to be single-arity?

**Comment by CrossEye** [comment]
`ap` is about compatibility with the [FantasyLand specification][fl].  Any similarity with `juxt` is just a helpful coincidence.

  [fl]: https://github.com/fantasyland/fantasy-land#apply

**Comment by samueller** [comment]
What an odd name for an algebraic JavaScript specification :). The reason I asked is that `juxt` is very useful at least in contrived exercises. Like getting the range of a set of inputs `juxt(min, max)` (not the most efficient way, but useful). `juxt` also lends itself well to pointfree function composition in many scenarios.

We teach kids as young as 8 years old JavaScript in a functional style (we later teach them Clojure as well). For our exercises and purposes, the functionality of `juxt` is used often. So I'd like to put a vote in to adding `juxt` to the Ramda library.

**Comment by buzzdecafe** [comment]
> We teach kids as young as 8 years old JavaScript in a functional style (we later teach them Clojure as well).

@samueller that's fantastic -- and you are using ramda to do this? IMO that is the best possible use of the library, which itself started as a research project.

**Comment by buzzdecafe** [comment]
incidentally, the PF version above doesn't work anymore because of API change to `useWith`. This is how it would be written today:

```js
var juxt = R.useWith(R.ap, [R.identity, R.of]);
```

**Comment by CrossEye** [comment]
> We teach kids as young as 8 years old JavaScript in a functional style (we later teach them Clojure as well).

:fireworks: :balloon: :cake: :star: :sparkling_heart: :sparkles:

**Comment by samueller** [comment]
@buzzdecafe yes we are working Ramda into our curriculum. We investigated all the other algebraic/functional JavaScript libraries and found Ramda most appealing. By far.

For supporting variadic functions, this should work:
```
var juxt = R.useWith(R.ap, [R.map(R.apply), R.of])
```
You would call it like this:
```
juxt([R.add, R.multiply], [3, 4]) //=> [7, 12]
```
